esp_custom_part_ota/src/esp_custom_part_ota.c

```diff
@@ -0,0 +1,281 @@
+        return ESP_ERR_INVALID_ARG;
+    }
+    esp_custom_part_ota_t *handle = (esp_custom_part_ota_t *)ctx;
+    if (handle->wrote_size == 0) {
+        return ESP_ERR_INVALID_ARG;
+    }
+    free(handle);
+    return ESP_OK;
+}
+
+esp_err_t esp_custom_part_ota_abort(esp_custom_part_ota_handle_t ctx)
+{
+    if (!ctx) {
+        return ESP_ERR_INVALID_ARG;
+    }
+    esp_custom_part_ota_t *handle = (esp_custom_part_ota_t *)ctx;
+    free(handle);
+    return ESP_OK;
+}
+
+esp_err_t esp_custom_part_ota_set_backup_partition(esp_custom_part_ota_handle_t ctx, const esp_partition_t *backup_partition)
+{
+    if (!ctx || !backup_partition) {
+        return ESP_ERR_INVALID_ARG;
+    }
+    esp_custom_part_ota_t *handle = (esp_custom_part_ota_t *)ctx;
+    handle->backup_partition = backup_partition;
+    return ESP_OK;
+}
+
+esp_err_t esp_custom_part_ota_partition_backup(esp_custom_part_ota_handle_t ctx, size_t backup_size)
+{
+    if (!ctx) {
+        return ESP_ERR_INVALID_ARG;
+    }
+    esp_custom_part_ota_t *handle = (esp_custom_part_ota_t *)ctx;
+    if (!handle->backup_partition) {
+        ESP_LOGE(TAG, "Backup partition not set. Cannot backup");
+        return ESP_FAIL;
+    }
+    handle->backup_len = (backup_size == 0) ? handle->update_partition->size : backup_size;
+    if (handle->backup_len > handle->backup_partition->size) {
+        ESP_LOGE(TAG, "Backup partition size smaller than data to be backed up");
+        return ESP_FAIL;
+    }
+    size_t block_size = 4096;
+    uint8_t *data = calloc(block_size, sizeof(uint8_t));
+    if (!data) {
+        ESP_LOGE(TAG, "Could not allocate memory for data");
+        return ESP_ERR_NO_MEM;
+    }
+
+    esp_err_t ret = esp_partition_erase_range(handle->backup_partition, 0, handle->backup_len);
+    if (ret != ESP_OK) {
+        ESP_LOGE(TAG, "Failed to erase the backup partition");
+        goto end;
+    }
+
+    if (handle->backup_len > 0) {
+        size_t copy_size = 0;
+        while (copy_size < handle->backup_len) {
+            size_t data_size = MIN(block_size, handle->backup_len - copy_size);
+            ret = esp_partition_read(handle->update_partition, copy_size, data, data_size);
+            if (ret != ESP_OK) {
+                goto end;
+            }
+            ret = esp_partition_write(handle->backup_partition, copy_size, data, data_size);
+            if (ret != ESP_OK) {
+                goto end;
+            }
+            copy_size += data_size;
+        }
+    }
+end:
+    free(data);
+    if (ret == ESP_OK) {
+        ret = esp_partition_erase_range(handle->update_partition, 0, handle->update_partition->size);
+        if (ret == ESP_OK) {
+            handle->need_erase = 0;
+            ESP_LOGI(TAG, "Successfully erased update partition");
+        }
+    }
+
+    // Store backup data information in the default NVS partition.
+    nvs_handle_t backup_info_handle = 0;
+    ret = nvs_open(BACKUP_STORAGE_NAMESPACE, NVS_READWRITE, &backup_info_handle);
```

Comment:
If the power goes at this moment then backup length is not yet stored and the primary partition is already erased. So this will result in data loss?

Reply:
Yes, it would have resulted in data loss. I have updated the code to first update the backup info in NVS and then erase the partition.